File: installer-cli/src/providers.rs
/// Custom provider configuration for unlisted providers.
#[derive(Debug, Clone)]
pub struct OtherProvider {
    pub provider: String,
    pub model: String,
    pub dimensions: Option<u32>, // Only for embeddings
    pub env_vars: Vec<(String, String)>,
}

/// Specification for a credential that needs to be collected.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CredentialSpec {
    pub env_var: &'static str,
    pub prompt: &'static str,
    pub secret: bool,
}

/// LLM provider selection for Cloud tier.
#[derive(Debug, Clone)]
pub enum LlmProvider {
    OpenAI,
    Anthropic,
    VertexAI,
    AzureOpenAI,
    Bedrock,
    Other(OtherProvider),
}

impl LlmProvider {
    pub fn required_credentials(&self) -> Vec<CredentialSpec> {
        match self {
            LlmProvider::OpenAI => vec![CredentialSpec {
                env_var: "OPENAI_API_KEY",
                prompt: "OpenAI API key",
                secret: true,
            }],
            LlmProvider::Anthropic => vec![CredentialSpec {
                env_var: "ANTHROPIC_API_KEY",
                prompt: "Anthropic API key",
                secret: true,
            }],
            LlmProvider::VertexAI => vec![
                CredentialSpec { env_var: "VERTEX_PROJECT", prompt: "GCP project ID", secret: false },
                CredentialSpec { env_var: "VERTEX_LOCATION", prompt: "GCP location (e.g., us-central1)", secret: false },
            ],
            LlmProvider::AzureOpenAI => vec![
                CredentialSpec { env_var: "AZURE_API_KEY", prompt: "Azure OpenAI API key", secret: true },
                CredentialSpec { env_var: "AZURE_API_BASE", prompt: "Azure endpoint URL", secret: false },
                CredentialSpec { env_var: "AZURE_API_VERSION", prompt: "Azure API version (e.g., 2024-02-01)", secret: false },
            ],
            LlmProvider::Bedrock => vec![
                CredentialSpec { env_var: "AWS_REGION", prompt: "AWS region (e.g., us-east-1)", secret: false },
            ],
            LlmProvider::Other(_) => vec![],
        }
    }

    /// litellm provider name for YAML output.
    pub fn provider_name(&self) -> &str {
        match self {
            LlmProvider::OpenAI => "openai",
            LlmProvider::Anthropic => "anthropic",
            LlmProvider::VertexAI => "vertex_ai",
            LlmProvider::AzureOpenAI => "azure",
            LlmProvider::Bedrock => "bedrock",
            LlmProvider::Other(o) => &o.provider,
        }
    }

    /// Default reasoning model for this provider.
    pub fn reasoning_model(&self) -> &str {
        match self {
            LlmProvider::OpenAI => "gpt-4o",
            LlmProvider::Anthropic => "claude-sonnet-4-5",
            LlmProvider::VertexAI => "gemini-2.5-pro",
            LlmProvider::AzureOpenAI => "gpt-4o",
            LlmProvider::Bedrock => "anthropic.claude-sonnet",
            LlmProvider::Other(o) => &o.model,
        }
    }

    /// Default fast model for this provider.
    pub fn fast_model(&self) -> &str {
        match self {
            LlmProvider::OpenAI => "gpt-4o-mini",
            LlmProvider::Anthropic => "claude-haiku-4-5",
            LlmProvider::VertexAI => "gemini-2.5-flash",
            LlmProvider::AzureOpenAI => "gpt-4o-mini",
            LlmProvider::Bedrock => "anthropic.claude-haiku",
            LlmProvider::Other(o) => &o.model,
        }
    }
}

/// Embedding provider selection for Cloud tier.
#[derive(Debug, Clone)]
pub enum EmbeddingProvider {
    OpenAI,
    VertexAI,
    AzureOpenAI,
    Bedrock,
    Other(OtherProvider),
}

impl EmbeddingProvider {
    pub fn provider_name(&self) -> &str {
        match self {
            EmbeddingProvider::OpenAI => "openai",
            EmbeddingProvider::VertexAI => "vertex_ai",
            EmbeddingProvider::AzureOpenAI => "azure",
            EmbeddingProvider::Bedrock => "bedrock",
            EmbeddingProvider::Other(o) => &o.provider,
        }
    }

    pub fn model(&self) -> &str {
        match self {
            EmbeddingProvider::OpenAI => "text-embedding-3-large",
            EmbeddingProvider::VertexAI => "text-embedding-005",
            EmbeddingProvider::AzureOpenAI => "text-embedding-3-large",
            EmbeddingProvider::Bedrock => "amazon.titan-embed-text-v2",
            EmbeddingProvider::Other(o) => &o.model,
        }
    }

    pub fn dimensions(&self) -> u32 {
        match self {
            EmbeddingProvider::OpenAI => 3072,
            EmbeddingProvider::VertexAI => 768,
            EmbeddingProvider::AzureOpenAI => 3072,
            EmbeddingProvider::Bedrock => 1024,
            EmbeddingProvider::Other(o) => o.dimensions.unwrap_or(768),
        }
    }

    pub fn required_credentials(&self) -> Vec<CredentialSpec> {
        match self {
            EmbeddingProvider::OpenAI => vec![CredentialSpec {
                env_var: "OPENAI_API_KEY",
                prompt: "OpenAI API key",
                secret: true,
            }],
            EmbeddingProvider::VertexAI => vec![
                CredentialSpec { env_var: "VERTEX_PROJECT", prompt: "GCP project ID", secret: false },
                CredentialSpec { env_var: "VERTEX_LOCATION", prompt: "GCP location", secret: false },
            ],
            EmbeddingProvider::AzureOpenAI => vec![
                CredentialSpec { env_var: "AZURE_API_KEY", prompt: "Azure OpenAI API key", secret: true },
                CredentialSpec { env_var: "AZURE_API_BASE", prompt: "Azure endpoint URL", secret: false },
                CredentialSpec { env_var: "AZURE_API_VERSION", prompt: "Azure API version", secret: false },
            ],
            EmbeddingProvider::Bedrock => vec![
                CredentialSpec { env_var: "AWS_REGION", prompt: "AWS region", secret: false },
            ],
            EmbeddingProvider::Other(_) => vec![],
        }
    }
}

/// Reranker provider selection for Cloud tier.
#[derive(Debug, Clone)]
pub enum RerankerProvider {
    LocalTei,  // Bundled TEI reranker (bge-reranker-v2-m3)
    None,
    Cohere,
    VertexAI,
    Other(OtherProvider),
}

impl RerankerProvider {
    pub fn provider_name(&self) -> Option<&str> {
        match self {
            RerankerProvider::LocalTei => Some("tei"),
            RerankerProvider::None => Option::None,
            RerankerProvider::Cohere => Some("cohere"),
            RerankerProvider::VertexAI => Some("vertex_ai"),
            RerankerProvider::Other(o) => Some(&o.provider),
        }
    }

    pub fn model(&self) -> Option<&str> {
        match self {
            RerankerProvider::LocalTei => Some("BAAI/bge-reranker-v2-m3"),
            RerankerProvider::None => Option::None,
            RerankerProvider::Cohere => Some("rerank-v3.5"),
            RerankerProvider::VertexAI => Some("semantic-ranker-default@latest"),
            RerankerProvider::Other(o) => Some(&o.model),
        }
    }

    pub fn is_none(&self) -> bool {
        matches!(self, RerankerProvider::None)
    }

    pub fn is_local(&self) -> bool {
        matches!(self, RerankerProvider::LocalTei)
    }

    pub fn required_credentials(&self) -> Vec<CredentialSpec> {
        match self {
            RerankerProvider::LocalTei => vec![], // No credentials needed
            RerankerProvider::None => vec![],
            RerankerProvider::Cohere => vec![CredentialSpec {
                env_var: "COHERE_API_KEY",
                prompt: "Cohere API key",
                secret: true,
            }],
            RerankerProvider::VertexAI => vec![
                CredentialSpec { env_var: "VERTEX_PROJECT", prompt: "GCP project ID", secret: false },
                CredentialSpec { env_var: "VERTEX_LOCATION", prompt: "GCP location", secret: false },
            ],
            RerankerProvider::Other(_) => vec![],
        }
    }
}

/// Collection of all three provider selections for credential deduplication.
#[derive(Debug, Clone)]
pub struct ProviderSet {
    pub llm: LlmProvider,
    pub embedding: EmbeddingProvider,
    pub reranker: RerankerProvider,
}

impl ProviderSet {
    /// Get deduplicated list of all required credentials.
    pub fn required_credentials(&self) -> Vec<CredentialSpec> {
        let mut creds = Vec::new();
        creds.extend(self.llm.required_credentials());
        creds.extend(self.embedding.required_credentials());
        creds.extend(self.reranker.required_credentials());

        // Deduplicate by env_var
        creds.sort_by(|a, b| a.env_var.cmp(b.env_var));
        creds.dedup_by(|a, b| a.env_var == b.env_var);
        creds
    }

    /// Get custom env vars from Other providers.
    pub fn custom_env_vars(&self) -> Vec<(String, String)> {
        let mut vars = Vec::new();
        if let LlmProvider::Other(o) = &self.llm {
            vars.extend(o.env_vars.clone());
        }
        if let EmbeddingProvider::Other(o) = &self.embedding {
            vars.extend(o.env_vars.clone());
        }
        if let RerankerProvider::Other(o) = &self.reranker {
            vars.extend(o.env_vars.clone());
        }
        vars
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn other_provider_can_store_env_vars() {
        let p = OtherProvider {
            provider: "groq".to_string(),
            model: "llama-3-70b".to_string(),
            dimensions: None,
            env_vars: vec![("GROQ_API_KEY".to_string(), "test".to_string())],
        };
        assert_eq!(p.env_vars.len(), 1);
        assert_eq!(p.env_vars[0].0, "GROQ_API_KEY");
    }

    #[test]
    fn llm_provider_names_are_lowercase() {
        assert_eq!(LlmProvider::OpenAI.provider_name(), "openai");
        assert_eq!(LlmProvider::Anthropic.provider_name(), "anthropic");
        assert_eq!(LlmProvider::VertexAI.provider_name(), "vertex_ai");
    }

    #[test]
    fn llm_provider_has_default_models() {
        assert_eq!(LlmProvider::OpenAI.reasoning_model(), "gpt-4o");
        assert_eq!(LlmProvider::OpenAI.fast_model(), "gpt-4o-mini");
        assert_eq!(LlmProvider::Anthropic.reasoning_model(), "claude-sonnet-4-5");
    }

    #[test]
    fn embedding_provider_has_dimensions() {
        assert_eq!(EmbeddingProvider::OpenAI.dimensions(), 3072);
        assert_eq!(EmbeddingProvider::VertexAI.dimensions(), 768);
        assert_eq!(EmbeddingProvider::Bedrock.dimensions(), 1024);
    }

    #[test]
    fn reranker_none_returns_none_for_model() {
        assert!(RerankerProvider::None.model().is_none());
        assert!(RerankerProvider::None.is_none());
    }

    #[test]
    fn reranker_cohere_has_model() {
        assert_eq!(RerankerProvider::Cohere.model(), Some("rerank-v3.5"));
        assert!(!RerankerProvider::Cohere.is_none());
    }

    #[test]
    fn provider_set_deduplicates_credentials() {
        let set = ProviderSet {
            llm: LlmProvider::VertexAI,
            embedding: EmbeddingProvider::VertexAI,
            reranker: RerankerProvider::VertexAI,
        };
        let creds = set.required_credentials();
        // Should have VERTEX_PROJECT and VERTEX_LOCATION only once each
        let vertex_project_count = creds.iter().filter(|c| c.env_var == "VERTEX_PROJECT").count();
        assert_eq!(vertex_project_count, 1);
    }

    #[test]
    fn provider_set_collects_mixed_credentials() {
        let set = ProviderSet {
            llm: LlmProvider::Anthropic,
            embedding: EmbeddingProvider::OpenAI,
            reranker: RerankerProvider::Cohere,
        };
        let creds = set.required_credentials();
        let env_vars: Vec<_> = creds.iter().map(|c| c.env_var).collect();
        assert!(env_vars.contains(&"ANTHROPIC_API_KEY"));
        assert!(env_vars.contains(&"OPENAI_API_KEY"));
        assert!(env_vars.contains(&"COHERE_API_KEY"));
    }
}

File: installer-go/internal/core/providers.go
package core

// EnvVar is a key-value pair for custom provider environment variables.
type EnvVar struct {
	Key   string
	Value string
}

// OtherProvider holds configuration for custom/unlisted providers.
type OtherProvider struct {
	Provider   string
	Model      string
	Dimensions *uint32
	EnvVars    []EnvVar
}

// CredentialSpec describes a credential that needs to be collected from the user.
type CredentialSpec struct {
	EnvVar string
	Prompt string
	Secret bool
}

// LlmProvider identifies an LLM provider.
type LlmProvider int

const (
	LlmOpenAI LlmProvider = iota
	LlmAnthropic
	LlmGeminiAPI
	LlmVertexAI
	LlmAzureOpenAI
	LlmBedrock
	LlmOther
)

// llmOtherData holds the OtherProvider payload when LlmProvider == LlmOther.
// Callers use NewLlmOther / LlmOtherData to work with it.
type llmProviderConfig struct {
	kind  LlmProvider
	other *OtherProvider
}

// LlmProviderConfig wraps an LlmProvider variant together with optional Other data.
type LlmProviderConfig struct {
	Kind  LlmProvider
	Other *OtherProvider
}

func NewLlmProvider(kind LlmProvider) LlmProviderConfig {
	return LlmProviderConfig{Kind: kind}
}

func NewLlmOther(o OtherProvider) LlmProviderConfig {
	return LlmProviderConfig{Kind: LlmOther, Other: &o}
}

func (p LlmProviderConfig) ProviderName() string {
	switch p.Kind {
	case LlmOpenAI:
		return "openai"
	case LlmAnthropic:
		return "anthropic"
	case LlmGeminiAPI:
		return "gemini"
	case LlmVertexAI:
		return "vertex_ai"
	case LlmAzureOpenAI:
		return "azure"
	case LlmBedrock:
		return "bedrock"
	case LlmOther:
		if p.Other != nil {
			return p.Other.Provider
		}
	}
	return ""
}

func (p LlmProviderConfig) ReasoningModel() string {
	switch p.Kind {
	case LlmOpenAI:
		return "gpt-4o"
	case LlmAnthropic:
		return "claude-sonnet-4-5"
	case LlmGeminiAPI:
		return "gemini-2.5-pro"
	case LlmVertexAI:
		return "gemini-2.5-pro"
	case LlmAzureOpenAI:
		return "gpt-4o"
	case LlmBedrock:
		return "anthropic.claude-sonnet"
	case LlmOther:
		if p.Other != nil {
			return p.Other.Model
		}
	}
	return ""
}

func (p LlmProviderConfig) FastModel() string {
	switch p.Kind {
	case LlmOpenAI:
		return "gpt-4o-mini"
	case LlmAnthropic:
		return "claude-haiku-4-5"
	case LlmGeminiAPI:
		return "gemini-2.5-flash"
	case LlmVertexAI:
		return "gemini-2.5-flash"
	case LlmAzureOpenAI:
		return "gpt-4o-mini"
	case LlmBedrock:
		return "anthropic.claude-haiku"
	case LlmOther:
		if p.Other != nil {
			return p.Other.Model
		}
	}
	return ""
}

func (p LlmProviderConfig) RequiredCredentials() []CredentialSpec {
	switch p.Kind {
	case LlmOpenAI:
		return []CredentialSpec{{EnvVar: "OPENAI_API_KEY", Prompt: "OpenAI API key", Secret: true}}
	case LlmAnthropic:
		return []CredentialSpec{{EnvVar: "ANTHROPIC_API_KEY", Prompt: "Anthropic API key", Secret: true}}
	case LlmGeminiAPI:
		return []CredentialSpec{{EnvVar: "GEMINI_API_KEY", Prompt: "Gemini API key (from ai.google.dev)", Secret: true}}
	case LlmVertexAI:
		return []CredentialSpec{
			{EnvVar: "VERTEX_PROJECT", Prompt: "GCP project ID", Secret: false},
			{EnvVar: "VERTEX_LOCATION", Prompt: "GCP location (e.g., us-central1)", Secret: false},
		}
	case LlmAzureOpenAI:
		return []CredentialSpec{
			{EnvVar: "AZURE_API_KEY", Prompt: "Azure OpenAI API key", Secret: true},
			{EnvVar: "AZURE_API_BASE", Prompt: "Azure endpoint URL", Secret: false},
			{EnvVar: "AZURE_API_VERSION", Prompt: "Azure API version (e.g., 2024-02-01)", Secret: false},
		}
	case LlmBedrock:
		return []CredentialSpec{{EnvVar: "AWS_REGION", Prompt: "AWS region (e.g., us-east-1)", Secret: false}}
	}
	return nil
}

// EmbeddingProvider identifies an embedding model provider.
type EmbeddingProvider int

const (
	EmbOpenAI EmbeddingProvider = iota
	EmbGeminiAPI
	EmbVertexAI
	EmbAzureOpenAI
	EmbBedrock
	EmbOther
)

// EmbeddingProviderConfig wraps an EmbeddingProvider variant with optional Other data.
type EmbeddingProviderConfig struct {
	Kind  EmbeddingProvider
	Other *OtherProvider
}

func NewEmbeddingProvider(kind EmbeddingProvider) EmbeddingProviderConfig {
	return EmbeddingProviderConfig{Kind: kind}
}

func NewEmbeddingOther(o OtherProvider) EmbeddingProviderConfig {
	return EmbeddingProviderConfig{Kind: EmbOther, Other: &o}
}

func (p EmbeddingProviderConfig) ProviderName() string {
	switch p.Kind {
	case EmbOpenAI:
		return "openai"
	case EmbGeminiAPI:
		return "gemini"
	case EmbVertexAI:
		return "vertex_ai"
	case EmbAzureOpenAI:
		return "azure"
	case EmbBedrock:
		return "bedrock"
	case EmbOther:
		if p.Other != nil {
			return p.Other.Provider
		}
	}
	return ""
}

func (p EmbeddingProviderConfig) Model() string {
	switch p.Kind {
	case EmbOpenAI:
		return "text-embedding-3-large"
	case EmbGeminiAPI:
		return "text-embedding-004"
	case EmbVertexAI:
		return "text-embedding-005"
	case EmbAzureOpenAI:
		return "text-embedding-3-large"
	case EmbBedrock:
		return "amazon.titan-embed-text-v2"
	case EmbOther:
		if p.Other != nil {
			return p.Other.Model
		}
	}
	return ""
}

func (p EmbeddingProviderConfig) Dimensions() uint32 {
	switch p.Kind {
	case EmbOpenAI:
		return 3072
	case EmbGeminiAPI:
		return 768
	case EmbVertexAI:
		return 768
	case EmbAzureOpenAI:
		return 3072
	case EmbBedrock:
		return 1024
	case EmbOther:
		if p.Other != nil && p.Other.Dimensions != nil {
			return *p.Other.Dimensions
		}
		return 768
	}
	return 768
}

func (p EmbeddingProviderConfig) RequiredCredentials() []CredentialSpec {
	switch p.Kind {
	case EmbOpenAI:
		return []CredentialSpec{{EnvVar: "OPENAI_API_KEY", Prompt: "OpenAI API key", Secret: true}}
	case EmbGeminiAPI:
		return []CredentialSpec{{EnvVar: "GEMINI_API_KEY", Prompt: "Gemini API key (from ai.google.dev)", Secret: true}}
	case EmbVertexAI:
		return []CredentialSpec{
			{EnvVar: "VERTEX_PROJECT", Prompt: "GCP project ID", Secret: false},
			{EnvVar: "VERTEX_LOCATION", Prompt: "GCP location", Secret: false},
		}
	case EmbAzureOpenAI:
		return []CredentialSpec{
			{EnvVar: "AZURE_API_KEY", Prompt: "Azure OpenAI API key", Secret: true},
			{EnvVar: "AZURE_API_BASE", Prompt: "Azure endpoint URL", Secret: false},
			{EnvVar: "AZURE_API_VERSION", Prompt: "Azure API version", Secret: false},
		}
	case EmbBedrock:
		return []CredentialSpec{{EnvVar: "AWS_REGION", Prompt: "AWS region", Secret: false}}
	}
	return nil
}

// RerankerProvider identifies a reranker provider.
type RerankerProvider int

const (
	RerankerLocalTeiMiniLM RerankerProvider = iota
	RerankerLocalTeiJina
	RerankerNone
	RerankerCohere
	RerankerVertexAI
	RerankerOther
)

// RerankerProviderConfig wraps a RerankerProvider variant with optional Other data.
type RerankerProviderConfig struct {
	Kind  RerankerProvider
	Other *OtherProvider
}

func NewRerankerProvider(kind RerankerProvider) RerankerProviderConfig {
	return RerankerProviderConfig{Kind: kind}
}

func NewRerankerOther(o OtherProvider) RerankerProviderConfig {
	return RerankerProviderConfig{Kind: RerankerOther, Other: &o}
}

func (p RerankerProviderConfig) ProviderName() (string, bool) {
	switch p.Kind {
	case RerankerLocalTeiMiniLM:
		return "tei", true
	case RerankerLocalTeiJina:
		return "tei", true
	case RerankerNone:
		return "", false
	case RerankerCohere:
		return "cohere", true
	case RerankerVertexAI:
		return "vertex_ai", true
	case RerankerOther:
		if p.Other != nil {
			return p.Other.Provider, true
		}
	}
	return "", false
}

func (p RerankerProviderConfig) Model() (string, bool) {
	switch p.Kind {
	case RerankerLocalTeiMiniLM:
		return "cross-encoder/ms-marco-MiniLM-L6-v2", true
	case RerankerLocalTeiJina:
		return "jinaai/jina-reranker-v2-base-multilingual", true
	case RerankerNone:
		return "", false
	case RerankerCohere:
		return "rerank-v3.5", true
	case RerankerVertexAI:
		return "semantic-ranker-default@latest", true
	case RerankerOther:
		if p.Other != nil {
			return p.Other.Model, true
		}
	}
	return "", false
}

func (p RerankerProviderConfig) IsNone() bool {
	return p.Kind == RerankerNone
}

func (p RerankerProviderConfig) IsLocal() bool {
	return p.Kind == RerankerLocalTeiMiniLM || p.Kind == RerankerLocalTeiJina
}

func (p RerankerProviderConfig) MemoryLimit() string {
	switch p.Kind {
	case RerankerLocalTeiMiniLM:
		return "1G"
	case RerankerLocalTeiJina:
		return "6G"
	}
	return "1G"
}

func (p RerankerProviderConfig) RequiredCredentials() []CredentialSpec {
	switch p.Kind {
	case RerankerCohere:
		return []CredentialSpec{{EnvVar: "COHERE_API_KEY", Prompt: "Cohere API key", Secret: true}}
	case RerankerVertexAI:
		return []CredentialSpec{
			{EnvVar: "VERTEX_PROJECT", Prompt: "GCP project ID", Secret: false},
			{EnvVar: "VERTEX_LOCATION", Prompt: "GCP location", Secret: false},
		}
	}
	return nil
}

// ProviderSet holds a complete selection of LLM, embedding, and reranker providers.
type ProviderSet struct {
	LLM       LlmProviderConfig
	Embedding EmbeddingProviderConfig
	Reranker  RerankerProviderConfig
}

// RequiredCredentials returns a deduplicated list of credentials needed by all providers.
func (ps ProviderSet) RequiredCredentials() []CredentialSpec {
	seen := make(map[string]bool)
	var result []CredentialSpec
	for _, cred := range append(append(
		ps.LLM.RequiredCredentials(),
		ps.Embedding.RequiredCredentials()...),
		ps.Reranker.RequiredCredentials()...) {
		if !seen[cred.EnvVar] {
			seen[cred.EnvVar] = true
			result = append(result, cred)
		}
	}
	return result
}

// CustomEnvVars returns environment variables from any Other providers in the set.
func (ps ProviderSet) CustomEnvVars() []EnvVar {
	var vars []EnvVar
	if ps.LLM.Kind == LlmOther && ps.LLM.Other != nil {
		vars = append(vars, ps.LLM.Other.EnvVars...)
	}
	if ps.Embedding.Kind == EmbOther && ps.Embedding.Other != nil {
		vars = append(vars, ps.Embedding.Other.EnvVars...)
	}
	if ps.Reranker.Kind == RerankerOther && ps.Reranker.Other != nil {
		vars = append(vars, ps.Reranker.Other.EnvVars...)
	}
	return vars
}
